feature(http): adds a service for signing and validating URLs

hypeJunction opened the pull request:
URLs can now be signed with a HMAC key and validated at a later stage using
elgg_http_get_signed_url() and elgg_http_validate_signed_url().

Fixes #9884

hypeJunction committed:
feature(http): adds a service for signing and validating URLs

URLs can now be signed with a HMAC key and validated at a later stage using
elgg_http_get_signed_url() and elgg_http_validate_signed_url().

Fixes #9884

hypeJunction commented:
@ewinslow This would make #8647 quite easy to implement.

hypeJunction commented:
@mrclay When you get a chance, I would like to hear your feedback on this. Any security implications that I missing?

mrclay commented on engine/classes/Elgg/Di/ServiceProvider.php:
```diff
@@ -60,6 +60,7 @@
  * @property-read \Elgg\Router                             $router
  * @property-read \Elgg\Application\ServeFileHandler       $serveFileHandler
  * @property-read \ElggSession                             $session
+ * @property-read \Elgg\Security\SignedUrl                 $signedUrl
```
UrlSigner / $urlSigner ?

mrclay commented:
LGTM, though it might make sense to migrate something to use this as a test. Like email validation or forgot password.

mrclay commented on engine/classes/Elgg/Security/SignedUrl.php:
```diff
@@ -0,0 +1,90 @@
+		$parts = parse_url($url);
+
+		if (isset($parts['query'])) {
+			$query = elgg_parse_str($parts['query']);
```
You might throw here if `__elgg_mac` is already set.

hypeJunction committed:
feat(passwords): strengthen change password link with a HMAC signature

Change password URL is now generated with a HMAC signature reducing the risk of
unathorized password reset

hypeJunction committed:
feat(uservalidationbyemail): validation URLs are now signed with a HMAC key

Links sent in an email are now signed with a HMAC key, instead of a custom
code sent with a URL query.

hypeJunction commented:
@mrclay, it didn't make sense getting rid of the code in the password reset because it's used in a form, so just added another layer there. Uservalidationbyemail now uses the new API.

mrclay commented on engine/classes/Elgg/PasswordService.php:
```diff
@@ -89,7 +89,8 @@ function sendNewPasswordRequest($user_guid) {
 
 		// generate link
 		$link = _elgg_services()->config->getSiteUrl() . "changepassword?u=$user_guid&c=$code";
```
We don't need the $code anymore.

hypeJunction replied:
I didn't want to hack my way through the password service. The code is injected in the password reset form and then validated. There would be BC implications removing it entirely.

mrclay commented on engine/lib/elgglib.php:
```diff
@@ -1293,6 +1293,41 @@ function elgg_http_url_is_identical($url1, $url2, $ignore_params = array('offset
+ * Issues 403 response if signature is inalid
+ * @return void
+ */
+function elgg_http_validate_request_signature() {
```
I assume a "validate" function will return a bool. Maybe `elgg_assert_valid_signed_request` or `elgg_signed_request_gatekeeper`?

hypeJunction replied:
I like `elgg_signed_url_gatekeeper()`

mrclay commented:
We might want to add caveats that signed URLs are **not bound to the session** and shouldn't be used for actions in general. E.g. We might be tempted to replace action URLs with these to avoid the hassle of validating the parameters, but this would allow a leaked URL to be used by others, and possibly replayed; the permissions system is a bit of a fail-safe here, but it isn't everything.

It *would* be safe, however, to sign full action URLs and use the signed URL gatekeeper function at the top of the action.

hypeJunction commented:
@mrclay See if this covers it.

mrclay commented:
LGTM

hypeJunction commented:
It's been two weeks, merging.

Wouter0100 commented:
Muh.. This change isn't really BC. You've just deleted the `c` input parameter instead of deprecated? Shouldn't these things get removed in a BC-breaking release? (e.g. 3.x, 4.x).

hypeJunction commented:
What does it break?

Wouter0100 commented:
The old verification method with the c parameter.

hypeJunction commented:
Why do you need the c parameter? URL is signed and verified.

hypeJunction commented:
It's internal plugin logic, and you shouldn't rely on system generated URLs to stay the same.

mrclay commented:
I'm sorry this hit you. It can be hard to identify the edges of the API, but we now lean towards progress over my paranoid conservatism.

Wouter0100 commented:
Hmm, yeah, I can understand it may be an edge case. In the case system generated URLs will change, we could still have it BC compatible, by providing a general method for creating system generated URLs (like possible with routes). For example, in [this](https://github.com/Centillien/validation_reminder) plugin we need to send an e-mail (different contents) with the validation URL (https://github.com/Centillien/validation_reminder/commit/4f460ab2132352472bca17d56bcd274d40d5f2bd), but due the change it wasn't working properly.

hypeJunction commented:
We just need to add a wrapper `uservalidationbyemail_get_validation_url()`

Wouter0100 commented:
@hypeJunction yep. Something like that.

hypeJunction commented:
Open an issue please.
